File: old/2021/rust/02.rs
fn solve(inp: &Vec<&str>) -> (i32, i32) {
    let (mut h, mut d) = (0, 0);
    let (mut ha, mut da, mut a) = (0, 0, 0);

    for i in 0..inp.len() {
        let values = inp[i].split(' ').collect::<Vec<_>>();

        match (values[0], values[1].parse::<i32>().unwrap()) {
            ("forward", n) => { h += n; ha += n; da += a * n; },
            ("down",    n) => { d += n; a += n; },
            ("up",      n) => { d -= n; a -= n; },
            _ => panic!(),
        }
    }


    (h * d, ha * da)
}

fn main() {
    let inp = std::fs::read_to_string("../input/02").unwrap();
    let inp = inp.trim().lines().collect::<Vec<_>>();

    let (part_one, part_two) = solve(&inp);

    println!("Part one: {}", part_one);
    println!("Part two: {}", part_two);
}

File: 2015/d02-i_was_told_there_would_be_no_math/langs/02.hs
module Main where

import Data.List
import Data.Char

get_sides :: [Char] -> Int -> [Int]
get_sides [] acc = acc : []
get_sides (x:xs) acc    | x == 'x' = acc : get_sides xs 0
                        | otherwise = get_sides xs ((ord x - 48) + acc * 10)

part_one :: [[Char]] -> Int
part_one [] = 0
part_one (x:xs) = case (sort (get_sides x 0)) of
    [l, w, h] -> 2 * l * w + 2 * w * h + 2 * h * l + l * w + part_one xs
    _ -> 0

part_two :: [[Char]] -> Int
part_two [] = 0
part_two (x:xs) = case (sort (get_sides x 0)) of
    [l, w, h] -> 2*l + 2*w + l*w*h + part_two xs
    _ -> 0

main :: IO ()
main = do
    inp <- readFile "../../../input/2015/02"
    let l = lines inp
    putStrLn ("Part one: " ++ (show (part_one l)))
    putStrLn ("Part two: " ++ (show (part_two l)))
